### PR TITLE
Handle edge case of a job mutating arguments

### LINKED ISSUE #221: Edge case: the handle_failure received a modified version of the job args
Here is the edge case.

Let's say you have a job failing and you are trying to enqueue it again in a "failed_job" queue or something similar.

What if that job is changing the value of an argument during it's processing, well, the handle_failure will receive a modified version of the job args.

Example:

```ruby
class MyJob
  def self.perform id, ids
    ids.map! { |id| id * 2 }
    1 / 0
  end
end
```

Now, you won't have the same values...but the ids * 2.

Obviously, this example is shitty because no one would do that, but there could be better reasons, like mapping ids to instances of classes. It doesn't sound like the best idea, but right now, the queue classic worker is crashing, which is really not cool!

Code of the process method is here: https://github.com/ryandotsmith/queue_classic/blob/master/lib/queue_classic/worker.rb#L111

What we could do is add as a first line that looks like this: `original_job = job.dup` and use that `original_job` when calling the handle_failure method.

I did not try it, I just thought about this now after debugging a similar issue.

How does that sound?

### PR DESCRIPTION
**Pull request by shosti**
The arguments to the job will be shallow-copied before passed to
handle_failure. Edge cases where nested data structures are mutated
are still possible (though probably very unlikely).

Fixes #221.

### CODE REVIEW

**Comment by senny** [comment]
@shosti thank you.

This breaks our test suite: `can't clone Fixnum`

**Comment by shosti** [comment]
OK, fixed it (I guess they changed `Fixnum#frozen?` between Ruby 1.9.3 and 2).

File: lib/queue_classic/worker.rb [hunk 1]
```diff
@@ -161,5 +163,15 @@ def setup_queues(adapter, queue, queues, top_bound)
       end
     end
 
+    def clone_args(job)
+      job[:args].map do |arg|
+        begin
+          arg.clone
+        rescue TypeError
```
Comment (jipiboily) [inline lib/queue_classic/worker.rb]: Why the rescue? What use case is it covering? I don't see anything in the tests covering that.

File: lib/queue_classic/worker.rb [hunk 2]
```diff
@@ -161,5 +163,15 @@ def setup_queues(adapter, queue, queues, top_bound)
       end
     end
 
+    def clone_args(job)
+      job[:args].map do |arg|
```
Comment (jipiboily) [inline lib/queue_classic/worker.rb]: Why is this a map? Could you just do (maybe I am being naive here!) `jobs[:args] = jobs[:args].clone`?

Comment (shosti) [inline lib/queue_classic/worker.rb]: `clone` does a shallow copy, so each individual arg wouldn't get copied (only the `arg` array).

Comment (jipiboily) [inline lib/queue_classic/worker.rb]: Makes sense, thanks!

File: test/worker_test.rb [hunk 3]
```diff
@@ -6,6 +6,10 @@ def no_args; return nil; end
   def one_arg(a); return a; end
   def two_args(a,b); return [a,b]; end
   def forty_two; OpenStruct.new(number: 42); end
+  def mutate_args(numbers)
+    numbers.map! {|number| number * 2 }
+    raise 'Problem!'
```
Comment (jipiboily) [inline test/worker_test.rb]: Why a `raise`?

**Comment by jipiboily** [comment]
Thanks for this!

This is something I had in mind for a little while, just never got to it :grinning: 

@senny: do you think we should merge that in 3.1 or wait to 3.2? Thoughts?

File: lib/queue_classic/worker.rb (hunk 1 above)
Comment (shosti) [inline lib/queue_classic/worker.rb]: It's covering `Fixnum`, which throws a `TypeError` when you try to clone it (there are a bunch of tests that cover this case already, although I could add another explicit one). I can't find another general way to find out if an object can be safely cloned—`frozen?` delivers `true` for `Fixnum`s in Ruby >=2.0.0 but `false` in <=1.9.3, and `respond_to?` returns `true` for `Fixnum` even though it blows up. The other option I guess is to specifically check for `Fixnum` or `Float`, which would probably be safe because it's serialized from JSON.

File: test/worker_test.rb (hunk 3 above)
Comment (shosti) [inline test/worker_test.rb]: Just simulating an error—I'll make the method name more explicit.

File: lib/queue_classic/worker.rb (hunk 1 above)
Comment (jipiboily) [inline lib/queue_classic/worker.rb]: Oh, thanks for the explanation!

I would love to see an explicit test for this unless it's hard to do, but it shouldn't from what I understand.

**Comment by senny** [comment]
@jipiboily from my point of view we shouldn't backport. While unlikely, someone might depend on this behavior. This changes the public API specifically `handle_failure`. Previously it was possible to pass arguments through that hash to `handle_failure`.

**Comment by jipiboily** [comment]
Certainly not backport. I guess you mean that you consider merging in 3.1 now a backport as an RC was released, correct?

**Comment by senny** [comment]
@jipiboily yup.

**Comment by jipiboily** [comment]
Agreed.

**Comment by shosti** [comment]
OK, PR updated with the extra test and some clarifications.

**Comment by jipiboily** [comment]
@shosti: looks good at first, going to re-review when I have more time but it's not going to be merged right away, I'll wait until 3.1 is out (a matter of weeks).

Thanks a lot!

**Comment by senny** [comment]
Thinking about this PR I'm not yet convinced cloning the arguments is necessary.

When you write a function, which accepts arguments mutating them is almost always a bad idea. You should make a copy you own before updating them. On the other hand there are certain cases where that mutation is the purpose. For example, consider a custom queue classic worker, which logs not only the jobs arguments but also a result (calculated by the job). Currently the job could accept a blank Hash `{}` and then use that as storage to pass results to the worker.

Looking at the implementation I further think this will leave us in an nonsatisfying state. We now try to clone the argument for the user to protect him from unwanted mutations, however since we don't know how to clone every object we fall back to passing to original. From a user perspective this still means that it's potentially dangerous to modify these objects.

I think we're probably better of giving the user a sharp tool which has the potential to cut yourself.

**Comment by shosti** [comment]
@senny One minor clarification: falling back to the original object should only happen for numbers, which can't be mutated anyways (all other JSON data types are cloneable).

**Comment by senny** [comment]
@shosti yea you'r right.

As a side note. In that case it'd be probably clear if we'd switch on the object type `Numeric` instead of rescuing `TypeError`.

@jipiboily I'll leave the final decision up to you. I don't feel to strongly about it but I'm fine with how things are at the moment (having side-effects when mutating args).

**Comment by jipiboily** [comment]
Closing for now as it was an internal problem that we faced only once. Also, I didn't think of the side effects @senny exposed.
